doghouse/appengine/github.go

```diff
@@ -0,0 +1,236 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/google/go-github/github"
+	"github.com/haya14busa/reviewdog/doghouse/server"
+	"github.com/haya14busa/reviewdog/doghouse/server/cookieman"
+	"golang.org/x/oauth2"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/log"
+	"google.golang.org/appengine/urlfetch"
+)
+
+type GitHubHandler struct {
+	clientID     string
+	clientSecret string
+
+	tokenStore     *cookieman.CookieStore
+	redirURLStore  *cookieman.CookieStore // Redirect URL after login.
+	authStateStore *cookieman.CookieStore
+
+	privateKey    []byte
+	integrationID int
+}
+
+func NewGitHubHandler(clientID, clientSecret string, c *cookieman.CookieMan, privateKey []byte, integrationID int) *GitHubHandler {
+	return &GitHubHandler{
+		clientID:       clientID,
+		clientSecret:   clientSecret,
+		tokenStore:     c.NewCookieStore("github-token", nil),
+		redirURLStore:  c.NewCookieStore("github-redirect-url", nil),
+		authStateStore: c.NewCookieStore("github-auth-state", nil),
+		integrationID:  integrationID,
+		privateKey:     privateKey,
+	}
+}
+
+func (g *GitHubHandler) buildGithubAuthURL(r *http.Request, state string) string {
+	var redirURL url.URL
+	redirURL = *r.URL
+	redirURL.Path = "/gh/_auth/callback"
+	redirURL.RawQuery = ""
+	redirURL.Fragment = ""
+	const baseURL = "https://github.com/login/oauth/authorize"
+	authURL := fmt.Sprintf("%s?client_id=%s&redirect_url=%s&state=%s",
+		baseURL, g.clientID, redirURL.RequestURI(), state)
+	return authURL
+}
+
+func (g *GitHubHandler) HandleAuthCallback(w http.ResponseWriter, r *http.Request) {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method GitHubHandler.HandleAuthCallback should have comment or be unexported